102 add md calcjob

federicazanca opened the pull request.

ElliottKasoar reviewed (commented):
Out of interest, are you able to run `--help` or something to list the options, and if so, if that affected by inheritance between the calculations, or is that side of things handled separately?

ElliottKasoar commented:
> Out of interest, are you able to run `--help` or something to list the options, and if so, if that affected by inheritance between the calculations, or is that side of things handled separately?

To clarify this a bit better, a few examples of shared arguments for typer are discussed [here](https://github.com/tiangolo/typer/issues/405).

With janus, for example, if I defined the shared `architecture` option via the callback method proposed in the issue above, the option would be listed under `janus --help`, but not under `janus singlepoint --help`, which is not ideal.

Clearly the CLI is built quite differently with click, classes, etc. here, but I can imagine an issue that looks similar could occur -  the help may not include inherited options.

It's probably fine, but if's something to check.

federicazanca commented:
Some things to do still
- [ ] make some sort of results_dict
- [ ] README file/docs
- [ ] Maybe I need some tests for the base scripts I created 
- [ ] check the --help thing Elliot said

But feel free to start commenting on the code as these thing won't modify (I think) what I've done so far

alinelena commented on aiida_mlip/calculations/base.py:
```diff
@@ -0,0 +1,222 @@
+            model_path = self.inputs.model.filepath
+        else:
+            model_path = ModelData.download(
+                "https://github.com/stfc/janus-core/raw/main/tests/models/mace_mp_small.model",  # pylint:disable=line-too-long
```
once psdi will have the repo in place... we can change to that

ElliottKasoar commented on aiida_mlip/calculations/base.py:
```diff
@@ -0,0 +1,222 @@
+
+        codeinfo = datastructures.CodeInfo()
+
+        # Initialize cmdline_params as an empty list
+        codeinfo.cmdline_params = []
+
+        # Adding command line params for when we run janus
+        codeinfo.cmdline_params.append("calculation")
```
```suggestion
        # Initialize cmdline_params
        codeinfo.cmdline_params = ["calculation"]

        # Adding command line params for when we run janus
```

ElliottKasoar commented on aiida_mlip/calculations/md.py:
```diff
@@ -0,0 +1,111 @@
+        """
+        super().define(spec)
+
+        # Additional inputs for molecula dynamics
```
```suggestion
        # Additional inputs for molecular dynamics
```

ElliottKasoar commented on aiida_mlip/calculations/md.py:
```diff
@@ -0,0 +1,111 @@
+
+        md_dictionary = self.inputs.md_dict.get_dict()
+
+        if not "traj-file" in md_dictionary:
+            md_dictionary["traj-file"] = str(self._DEFAULT_TRAJ_FILE)
+        if not "stats-file" in md_dictionary:
+            md_dictionary["stats-file"] = str(self._DEFAULT_STATS_FILE)
```
```suggestion
        md_dictionary.setdefault("traj-file", str(self._DEFAULT_TRAJ_FILE))
        md_dictionary.setdefault("stats-file", str(self._DEFAULT_STATS_FILE))
```

ElliottKasoar commented on aiida_mlip/calculations/singlepoint.py:
```diff
@@ -176,70 +111,17 @@ def prepare_for_submission(
-            self.uuid,
-            log_filename,
-        ]
+        codeinfo.cmdline_params[0] = "singlepoint"
```
Is this just overwriting the placeholder(?) "calculation"?

Might be worth adding a comment (here and elsewhere), as otherwise it's not immediately obvious.

ElliottKasoar commented on README.md:
```diff
@@ -43,17 +43,20 @@ Registered entry points for aiida.calculations:
-verdi run submit_singlepoint.py "janus@localhost" --architecture mace --model "/path/to/model"    # run test calculation
-verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully-opt True
+verdi run submit_singlepoint.py "janus@localhost" --architecture mace --model "/path/to/model"    # run singlepoint calculation
+verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully-opt True # run geometry optimisation
```
Just to check, you have to pass `True`, not just `--fully-opt`?

federicazanca replied:
yes, I could try and change it? I like it better this way but maybe we want consistency? also this one is the example script anyway so maybe we don't care

ElliottKasoar commented:
Just a note that there will be minor changes to the janus CLI once https://github.com/stfc/janus-core/pull/100 is merged to allow reading a config.

The currently linked janus-core will be fine, but if we updated that before merging this, we'd need to take it into account.

In most cases I changed the Python variables to match the CLI option names, so the CLI is largely unchanged, but there is at least one difference: `--property` -> `--properties` (since `property` is a Python keyword)

ElliottKasoar commented on README.md:
```diff
@@ -54,7 +54,7 @@ A quick demo of how to submit a calculation using the provided example files:
 cd examples/calculations
 verdi run submit_singlepoint.py "janus@localhost" --architecture mace --model "/path/to/model"    # run singlepoint calculation
-verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully-opt True # run geometry optimisation
+verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully_opt True # run geometry optimisation
```
A similar note to before: using underscores differs from the default for click/typer option naming (see `Parameters` [here](https://click.palletsprojects.com/en/8.1.x/api/)) (and so also from janus currently). Are you sure we want this?

federicazanca replied:
I know, this was something I was pointing out when I asked you about the underscore. basically if I put fully-opt rather than fully_opt I get an error that it does not like it, maybe because of the way aiida itself is written (?), so I decided to make it with the _ also in click. Still can change it tho

ElliottKasoar replied:
[Random example](https://aiida-quantumespresso.readthedocs.io/en/latest/reference/cli/index.html#cmdoption-aiida-quantumespresso-calculation-launch-cp-F), but looks like it should work ok. We can have a look tomorrow maybe

ElliottKasoar commented on aiida_mlip/parsers/md_parser.py:
```diff
@@ -0,0 +1,119 @@
+            traj_filepath = md_dictionary.get("traj-file", MD.DEFAULT_TRAJ_FILE)
+            with self.retrieved.open(traj_filepath, "rb") as handle:
+                self.out("traj_file", SinglefileData(file=handle))
+            fin, traj_output = xyz_to_aiida_traj(
```
I'm not sure about `fin` as a variable name. Maybe `final_struct`?

federicazanca commented:
This PR also closes Issue #103

federicazanca commented:
I still need to write the tests for the config and documentation so some things might not be clear.
In the examples submit_using_config it is shown how to use the config file, written for janus, to submit a calculation. It is basically the same as before just the config file is a new parameter.

There is also the function run_from_config() in the base class, which should instead run aiida using a config file written for aiida. As discussed, this could be something that is split in 2 parts, 1 part for janus and one for the parameters specific to aiida. At the moment it does not work like this though. I can either remove it at all and we only keep that one way to use the config file (which probably makes sense, because running aiida using a config file should be implemented in aiida-core not here, it might even already exist), or I fix it to read a config file 'split' in 2 parts, one for aiida, one for janus.

ElliottKasoar commented:
> There is also the function run_from_config() in the base class, which should instead run aiida using a config file written for aiida. As discussed, this could be something that is split in 2 parts, 1 part for janus and one for the parameters specific to aiida. At the moment it does not work like this though. I can either remove it at all and we only keep that one way to use the config file (which probably makes sense, because running aiida using a config file should be implemented in aiida-core not here, it might even already exist), or I fix it to read a config file 'split' in 2 parts, one for aiida, one for janus.

Sorry, just to check, how does it currently work then? Does it only contain aiida config options (which is fine for now)?

ElliottKasoar commented on README.md:
```diff
@@ -43,17 +43,20 @@ Registered entry points for aiida.calculations:
-verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully-opt True
+verdi run submit_singlepoint.py "janus@localhost" --architecture mace --model "/path/to/model"    # run singlepoint calculation
+verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully_opt True # run geometry optimisation
+verdi run submit_md.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --ensemble "nve" --md_dict_str "'temp':300,'steps':4,'traj-every':3,'stats-every':1}" # run molecular dynamics
```
```suggestion
verdi run submit_md.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --ensemble "nve" --md_dict_str "{'temp':300,'steps':4,'traj-every':3,'stats-every':1}" # run molecular dynamics
```

federicazanca commented:
Summary of PR since it's gotten quite long

- **MD Calculation Addition**: #102 
  - Added MD calculation functionality.
  - Created a base class for calculations, with MD and single-point calculations inheriting from it (the code of the base should be clean-ish in theory since it mostly come from the original singlepoint.py)
  - Relevant files:
    - `/aiida_mlip/calculations/base.py`
    - `/aiida_mlip/calculations/md.py`
    - `/aiida_mlip/parsers/md_parser.py`
    - `/aiida_mlip/parsers/base_parser.py`
    - some changes also in the opt and sp files

- **Config File Option**: #109 
  - Implemented an option for passing arguments via a config file.
  - Added the "config" option in `base.py`.
  -New data type specifically for the config file (`config.py` in data). Not sure if it is REALLY needed or if the code the way it's wirtten is good at all, but I found it helpful to organise the functions connected to the config. 
  - The usage of the config file is shown in the examples:
    - `aiida_mlip/examples/calculations/submit_using_config.py`
    - `aiida_mlip/examples/calculations/submit_md_using_config.py`

- **New Helper Functions**: #103 
  - Added new helper functions/modified a bit some old ones.
  - Files:
    - `aiida_mlip/helpers/converter.py`
    - `aiida_mlip/helpers/converter.py`
  - Tests added for some of them.

ElliottKasoar commented on examples/calculations/submit_using_config.py:
```diff
@@ -0,0 +1,35 @@
+from aiida_mlip.data.config import JanusConfigfile
+from aiida_mlip.helpers.help_load import load_structure
+
+# And the required inputs for aiida
```
```suggestion
# Add the required inputs for aiida
```

federicazanca commented on aiida_mlip/calculations/base.py:
```diff
@@ -0,0 +1,250 @@
+        # Wrapping processes may choose to exclude certain input ports
+        # If the ports have been excluded, skip the validation.
+        if "struct" not in port_namespace and "config" not in port_namespace:
+            raise ValueError("'Structure' namespaces is required.")
```
this needs fixing in general caus I should also check inside the config if a structure is given. but I assume janus also raises an error if you don't give a structure, so maybe I can just update the error with 
"at least one between struct or config should be given"

ElliottKasoar replied:
Janus would throw an error, but you'd possibly hit other errors first setting up the nodes?

But that's a good point, just checking for `struct` or `config` could still mean you're missing the structure, if it's not in the config. Since it's quite fundamental, you probably want to handle it and not rely on it filtering through and being handled nicely by janus.

I'm not completely clear how the namespaces fit in, but something along the lines of "structure must be specified through struct or config" (with appropriate formatting/language for the namespaces) is probably what you want here, then another check in the config that it contains the structure if you don't have it already?

ElliottKasoar commented on examples/calculations/submit_md_using_config.py:
```diff
@@ -0,0 +1,31 @@
+
+# Run calculation
+result, node = run_get_node(
+    MDCalculation, code=code, struct=structure, metadata=metadata, config=config
```
It might be nice to add (here or the singlepoint example) explicit inputs as well as the config, so it's clear both are possible?

federicazanca replied:
Technically, `structure`, `code`, and `metadata` are explicit inputs. `Code` and `metadata` are just for aiida, so that's not helpful but `struct` is for janus (and actually at the moment in that config file there is a different structure used, but thet struct I pass here overrides it)

federicazanca replied:
also as a reminder for myself, I need to change the paths in the config and the submission files cause they are path in my localcomputer that I am just using to see if things work

federicazanca commented on aiida_mlip/calculations/md.py:
```diff
@@ -0,0 +1,128 @@
+        elif "ensemble" in self.inputs.config.as_dictionary.items():
+            ensemble = self.inputs.config.as_dictionary["ensemble"]
+        else:
+            print("ERROR ENSEMBLE NOT PROVIDED")
```
this one was for debugging, I forgot to remove it. I think rather than this I will put some input validation to ensure that somewhere the ensemble is given, does it sound better?
but if janus has a default for the ensemble I'll just make it use the default

ElliottKasoar replied:
There's no default at the moment, although it's something we have briefly considered. For now, yes catching it around here would be nice

federicazanca commented on aiida_mlip/data/config.py:
```diff
@@ -0,0 +1,148 @@
+        Parameters
+        ----------
+        store_all : bool
+            Define if you want to store all the parameters or only the main ones.
```
good point.
at the moment I am saving:
code, structure, model, architecture, ensemble and fully_opt, I can add others if we want.
this means that when you run aiida-mlip using the config file, only this parameters will be saved in the data provenance as single parameters. still all the other info in the config file are saved in the config file node, so they are not lost, they are just grouped under configfile basically.

federicazanca replied:
the fact that there is the option to choose to store all, is not a real option tbh. I put it there for the implementation of the config for aiida itself, but now that I removed it I could probably remove also this option and keep it simple

federicazanca replied:
ways this could be improved:
-give people a choice on which parameters they want to save singularly (I don't think we should, tho, it gives space to too many errors)
-explain in the documentation or somewhere that when using the config file only some parameters are saved individually, the rest can be found in config (which in the end is the same as passing a list of parameters as a dictionary so I think it's fine)

ElliottKasoar replied:
Yeah I think we probably want a set way that things are saved, outside of user control. There's probably not a particularly nice way to do anything else anyway.

So yes, I think the main thing is just making it clear (in the Python and the docs) what is saved differently

federicazanca commented on aiida_mlip/helpers/help_load.py:
```diff
@@ -66,7 +74,7 @@ def load_structure(struct: Optional[Union[str, Path, int]]) -> StructureData:
         If the input is not a valid path to a structure file or a node PK.
     """
-    if struct is None:
+    if struct is None or struct == "NaCl":
```
yea I know, it's horrible. basically in the load_structure function the default is to load NaCl, and I thought having struct = load_structure(None) looked bad so I added the option to call it using NaCl instead, I'll check what will happen if I have a file with that name

federicazanca commented on aiida_mlip/helpers/converters.py:
```diff
@@ -0,0 +1,105 @@
+        elif key == "struct":
+            value = load_structure(value)
+        elif key == "model":
+            value = load_model(value, arch)
+        elif key == "arch":
+            arch = value
```
it is and I'll try to change it to something better yea

oerc0122 commented on aiida_mlip/parsers/md_parser.py:
```diff
@@ -0,0 +1,116 @@
+        """
+        # Call the parent parse method to handle common parsing logic
+        exit_code = super().parse(**kwargs)
+        if exit_code == ExitCode(0):
+
```
This would probably be better as:
```suggestion
        if exit_code != ExitCode(0):
            return exit_code
```
And de-indent the rest of the function

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -1,9 +1,15 @@
 ==============================
 
-SinglePoint calculation
+In these descriptions, we will assume that the code `janus-core <https://github.com/stfc/janus-core>`_ is installed and saved in the AiiDA database as an InstalledCode instance named 'janus@localhost'.
```
```suggestion
In these examples, we will assume that the `janus-core <https://github.com/stfc/janus-core>`_ package is installed and saved in the AiiDA database as an `InstalledCode` instance named 'janus@localhost'.
```

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -1,9 +1,15 @@
-SinglePoint calculation
+In these descriptions, we will assume that the code `janus-core <https://github.com/stfc/janus-core>`_ is installed and saved in the AiiDA database as an InstalledCode instance named 'janus@localhost'.
+
+The structure file on which to run a calculation can be anything that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read, indicated with the path `path/to/structure`.
```
What is indicated? Do you mean "It is specified by"?
```suggestion
The structure should be a path to a file. Here, the structure file is specified as `path/to/structure`.

.. note::
   Any format that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read is a valid structure file for a calculation.
```

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -1,9 +1,15 @@
+
+The structure file on which to run a calculation can be anything that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read, indicated with the path `path/to/structure`.
+
+The model file is a local file or a file to download, depending on the specific MLIP that is being used. In these examples, it is referred to as a local file `path/to/model`.
```
```suggestion
The model file determines the specific MLIP to be used. I can be a local file or a URI to a file to download. In these examples, it is assumed to be a local file located at `path/to/model`.
```

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -18,24 +24,76 @@ Below is a usage example with the minimum required parameters. These parameters
 .. code-block:: python
 
+    SinglePointCalculation = CalculationFactory("janus.sp")
+    submit(SinglePointCalculation, code=InstalledCode, structure=StructureData, metadata="{"options": {"resources": {"num_machines": 1}}}")
```
This doesn't look right. Should those outer quotes be there?

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -57,6 +115,8 @@ Below is a usage example with some additional geometry optimisation parameters.
     submit(GeomOptCalculation, code=InstalledCode, structure=StructureData, max_force=Float(0.1), vectors_only=Bool(True))
 
 
+As per the singlepoint calculation, the parameters can be provided in the form of a dictionary or a config file.
```
```suggestion
.. note::

    As per the singlepoint calculation, the parameters can be provided as a dictionary or config file.
```

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -67,3 +127,37 @@ They will be converted to AiiDA data types by the script itself.
+Molecular Dynamics calculation
+------------------------------
+
+A `MD` Calculation represents a `Calcjob` object within the AiiDA framework.
```
```suggestion
An `MD` Calculation represents a `Calcjob` object within the AiiDA framework.
```

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -67,3 +127,37 @@ They will be converted to AiiDA data types by the script itself.
+
+
+    MDCalculation = CalculationFactory("janus.md")
+    submit(MDCalculation, code=InstalledCode, structure=StructureData, ensemble=Str("nve") md_dict=Dict({'temp':300,'steps': 4,'traj-every':3,'stats-every':1}))
```
```suggestion
    submit(MDCalculation, code=InstalledCode, structure=StructureData, ensemble=Str("nve"), md_dict=Dict({'temp':300,'steps': 4,'traj-every':3,'stats-every':1}))
```

oerc0122 commented on docs/source/user_guide/calculations.rst:
```diff
@@ -67,3 +127,37 @@ They will be converted to AiiDA data types by the script itself.
+    MDCalculation = CalculationFactory("janus.md")
+    submit(MDCalculation, code=InstalledCode, structure=StructureData, ensemble=Str("nve") md_dict=Dict({'temp':300,'steps': 4,'traj-every':3,'stats-every':1}))
+
+As per the singlepoint calculation, the parameters can be provided in the form of a dictionary or a config file.
```
```suggestion
.. note::

   As per the singlepoint calculation, the parameters can be provided as a dictionary or config file.
```

oerc0122 commented on docs/source/user_guide/data.rst:
```diff
@@ -30,3 +30,90 @@ Usage
 
     model = ModelData.download('http://yoururl.test/model', architecture='mace', filename='model', cache_dir='/home/mlip/', force_download=False)
+
+-The architecture of the model file can be accessed using the `architecture` property:
```
```suggestion
- The architecture of the model file can be accessed using the `architecture` property:
```

oerc0122 commented on docs/source/user_guide/data.rst:
```diff
@@ -30,3 +30,90 @@ Usage
+
+
+
+-The filepath of the model file can be accessed using the `filepath` property:
```
```suggestion
- The filepath of the model file can be accessed using the `filepath` property:
```

oerc0122 commented on docs/source/user_guide/data.rst:
```diff
@@ -30,3 +30,90 @@ Usage
+
+.. warning::
+
+    When sharing data, using the ``filepath`` could point to a location inaccessible on another computer.
```
```suggestion
    When using shared data, the ``filepath`` could point to a inaccessible location on another computer.
```

oerc0122 commented on docs/source/user_guide/data.rst:
```diff
@@ -30,3 +30,90 @@ Usage
+.. warning::
+
+    When sharing data, using the ``filepath`` could point to a location inaccessible on another computer.
+    So if you are using data from someone else, for both the modeldata and the configfile, consider using the ``get_content()`` method to create a new file with identical content.
```
```suggestion
    So if you are using data from someone else, for both the model data and the config file, consider using the ``get_content()`` method to create a new file with identical content.
```

oerc0122 commented on docs/source/user_guide/data.rst:
```diff
@@ -30,3 +30,90 @@ Usage
+
+.. code-block:: python
+
+    config_file.store_content(store_all: False, skip: [])
```
Are colons the right syntax here? This looks more like Groovy

oerc0122 commented on docs/source/user_guide/data.rst:
```diff
@@ -30,3 +30,90 @@ Usage
+    A more robust solution to this problem is going to be implemented.
+
+
+JanusConfigfile
```
Wondering if this documentation should be here or in the generated docs and docstrings with just a simple link to the API docs here

federicazanca replied:
I wonder the same, and I can (should?) add more info in the docstrings for the apidoc.
However, as a beginner/lazy user with little programming experience, I do remember finding api documentations confusing and preferring when everything was also written down in some user guide in a format with some examples.

If you are referring more specifically to the explanations for  `store_all` and `skip` that should definitely go in the api docs, maybe in a shorter format,  I put it here because it seemed a bit long.

oerc0122 replied:
Probably create an issue and get this in

oerc0122 commented on examples/calculations/submit_geomopt.py:
```diff
@@ -68,16 +66,39 @@ def geomopt(params: dict) -> None:
+    "--fully_opt",
+    default=False,
+    type=bool,
+    help="Fully optimize the cell vectors, angles, and atomic positions.",
```
You're using "optimisation" above, pick one.

oerc0122 commented on examples/calculations/submit_md_using_config.py:
```diff
@@ -0,0 +1,33 @@
+metadata = {"options": {"resources": {"num_machines": 1}}}
+code = load_code("janus@localhost")
+
+# # This structure will overwrite the one in the config file if present
```
```suggestion
# This structure will overwrite the one in the config file if present
```

oerc0122 commented on examples/calculations/submit_md_using_config.py:
```diff
@@ -0,0 +1,33 @@
+
+# # This structure will overwrite the one in the config file if present
+structure = load_structure(
+    "/home/federica/aiida-mlip/tests/calculations/structures/NaCl.cif"
```
Do you want these to be "/home/federica"?

oerc0122 commented on tests/calculations/test_geomopt.py:
```diff
@@ -62,16 +62,17 @@ def test_geomopt(fixture_sandbox, generate_calc_job, janus_code, model_folder):
 
-    print(calc_info.codes_info[0].cmdline_params)
-    print(cmdline_params)
+    print(sorted(map(str, calc_info.codes_info[0].cmdline_params)))
+    print(sorted(map(str, cmdline_params)))
```
Do we want printing during tests?

oerc0122 commented on tests/calculations/test_md.py:
```diff
@@ -0,0 +1,209 @@
+
+    calc_info = generate_calc_job(fixture_sandbox, entry_point_name, inputs)
+
+    cmdline_params = [
+        "md",
+        "--arch",
+        "mace",
+        "--struct",
+        "aiida.xyz",
+        "--device",
+        "cpu",
+        "--log",
+        "aiida.log",
+        "--summary",
+        "md_summary.yml",
+        "--calc-kwargs",
+        f"{{'default_dtype': 'float64', 'model': '{model_file}'}}",
+        "--ensemble",
+        "nve",
+        "--temp",
+        300.0,
+        "--steps",
+        4,
+        "--traj-every",
+        1,
+        "--stats-every",
+        1,
+        "--restart-every",
+        3,
+        "--traj-file",
+        "aiida-traj.xyz",
+        "--stats-file",
+        "aiida-stats.dat",
+    ]
```
Would it be better to list these as pairs? Or does black complain?

federicazanca replied:
it complains

oerc0122 commented on tests/calculations/test_md.py:
```diff
@@ -0,0 +1,209 @@
+    print(sorted(map(str, calc_info.codes_info[0].cmdline_params)))
+    print(sorted(map(str, cmdline_params)))
+    # Check the attributes of the returned `CalcInfo`
+    assert sorted(fixture_sandbox.get_content_list()) == ["aiida.xyz"]
```
I know its consistent, but sorting 1 element seems excessive.

oerc0122 commented on tests/conftest.py:
```diff
@@ -216,3 +216,25 @@ def model_folder():
         Path: The path to the example file.
     """
     return Path(__file__).resolve().parent / "data" / "input_files" / "mace"
```
May want to make 
```
    Path(__file__).resolve().parent
```
a constant

federicazanca commented on aiida_mlip/calculations/base.py:
```diff
@@ -0,0 +1,276 @@
+    # Wrapping processes may choose to exclude certain input ports
+    # If the ports have been excluded, skip the validation.
+
+    if not "struct" in inputs and not "config" in inputs:
```
it will use struct and not the one in config (which is also what janus does, maybe I should clarify it in the documentation?)

oerc0122 replied:
I would probably prefer an error raised rather than the situation being silently ignored.

ElliottKasoar replied:
I'm not sure I'd consider it being silently ignored, unless I'm misunderstanding. 

Firstly, the config can contain options other than structure, so until we read it later, we won't know at this point whether structure is duplicated or not. If we do want to address this, I think we'd need to do it in the config reader rather than here.

However, in discussions so far we've generally agreed that it can be useful for the config file to be overridable by explicit options, so if the structure is specified in both, using the explicit option is the expected behaviour. As @federicazanca mentioned, this is currently what janus does, but it's open to discussion.

oerc0122 replied:
You're probably right, I'm just reading this out of context. I guess at the very least it might be worth issuing a log message/warning at the point we do know.

federicazanca commented on aiida_mlip/calculations/md.py:
```diff
@@ -0,0 +1,128 @@
+        Create the input files for the `CalcJob`.
+    """
+
+    DEFAULT_TRAJ_FILE = "aiida-traj.xyz"
+    DEFAULT_STATS_FILE = "aiida-stats.dat"
+    DEFAULT_SUMMARY_FILE = "md_summary.yml"
```
I had them private but then actually I need to use them in the parser so I removed the `_`
But because I didn't use all of them in the parser those that I wasn not using I left them private, which I think is not nice and consistent, should I just change them all to not private?

oerc0122 replied:
Letting a power-user see and overwrite these is sensible. Not private is probably best.

federicazanca commented on aiida_mlip/calculations/singlepoint.py:
```diff
@@ -152,12 +95,12 @@ def validate_inputs(
         if "input_filename" in inputs:
             if not inputs["input_filename"].value.endswith(".xyz"):
-                raise ValueError("The parameter 'input_filename' must end with '.cif'")
+                raise ValueError("The parameter 'input_filename' must end with '.xyz'")
```
it is because of ase, I just wanted to catch the error at an aiida level so I can do stuff with it (like I am calling it Inputvalidationerror, rather than a generic valueerror).
but can remove it if it's redundant

federicazanca commented on aiida_mlip/calculations/singlepoint.py:
```diff
@@ -176,70 +119,18 @@ def prepare_for_submission(
-            self.uuid,
-            log_filename,
-        ]
+        # Singlepoint is overwriting the placeholder "calculation" from the base.py file
```
but they are the same? or do you mean this + the comment before this:
```
# Adding command line params for when we run janus
```

oerc0122 replied:
When scanning for some reason I read this one differently to the first. I don't know why. We can ignore this.

federicazanca commented on aiida_mlip/calculations/base.py:
```diff
@@ -0,0 +1,276 @@
+            config_dict = self.inputs.config.as_dictionary
+            overlapping_params = set(cmd_line.keys()) & set(config_dict.keys())
+            # Store the other parameters
+            self.inputs.config.store_content(skip=list(overlapping_params))
```
no, removed it

federicazanca commented on aiida_mlip/helpers/help_load.py:
```diff
@@ -38,11 +44,13 @@ def load_model(model: Optional[Union[str, Path]], architecture: str) -> ModelDat
 
 
-def load_structure(struct: Optional[Union[str, Path, int]]) -> StructureData:
+def load_structure(struct: Union[str, Path, int] = None) -> StructureData:
```
thing is, when I don't pass anything and I call it like
```
struct = load_structure()
```
it will consider that struct does not exist and won't work, so I now just kept both Optional and the default value None. If it makes sense (there might be a better solution)

oerc0122 replied:
`Optional` just means it accepts a `Nonetype`, so you can have either
```suggestion
def load_structure(struct: Optional[Union[str, Path, int]] = None) -> StructureData:
```
or
```suggestion
def load_structure(struct: Union[str, Path, int, None] = None) -> StructureData:
```

federicazanca commented on aiida_mlip/calculations/base.py:
```diff
@@ -0,0 +1,276 @@
+        raise InputValidationError(
+            "Structure must be specified through struct or config"
+        )
+    if "config" in inputs and "struct" not in inputs["config"].as_dictionary:
```
inputs["config"] is a Janusconfigfile data type that I created, it needs to be a dictionary or maybe a text to check if it contains struct

oerc0122 replied:
You could probably add a [`__contains__`](https://docs.python.org/3/reference/datamodel.html#object.__contains__) dunder method

oerc0122 commented on aiida_mlip/data/config.py:
```diff
@@ -0,0 +1,167 @@
+        """
+        return self.base.attributes.get("filepath")
+
+    @property
```
One other minor thing: depending on how big the file is and whether the file is liable to change, you may want to make this or `read_yaml` a [`cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) to reduce file-opens, but not important.

federicazanca replied:
I think it might be useful in general also for other things, but at the moment for this specifically it is not too important as the files won't be very big anyway. I'll put it in an issue

oerc0122 replied:
Agreed

oerc0122 commented on aiida_mlip/calculations/geomopt.py:
```diff
@@ -117,16 +112,28 @@ def prepare_for_submission(
+        if "vectors_only" in self.inputs:
+            geom_opt_cmdline["vectors-only"] = self.inputs.vectors_only.value
+        if "fmax" in self.inputs:
+            print()
```
Debugging statement?

federicazanca replied:
ops I'll remove all the prints now

oerc0122 commented on aiida_mlip/calculations/geomopt.py:
```diff
@@ -117,16 +112,28 @@ def prepare_for_submission(
-            "opt-kwargs": opt_kwargs,
-        }
+        geom_opt_cmdline = {"traj": self.inputs.traj.value}
+        if "opt_kwargs" in self.inputs:
```
What happens if they aren't? Are defaults handled in the called routines?

federicazanca replied:
if they aren't in inputs they might be in the config, if they are not in the config either it will use whatever default janus has for them.
I used to have the default in aiida too (that anyways it matched janus one) but I removed it because otherwise it would use the default and override the config file. We only want that when we intentionally set an input

ElliottKasoar commented on README.md:
```diff
@@ -43,17 +43,20 @@ Registered entry points for aiida.calculations:
 cd examples/calculations
-verdi run submit_singlepoint.py "janus@localhost" --architecture mace --model "/path/to/model"    # run test calculation
-verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --steps 5 --fully-opt True
+verdi run submit_singlepoint.py "janus@localhost" --architecture mace --model "/path/to/model"    # run singlepoint calculation
```
Does this not need a structure?

federicazanca replied:
In theory there are default values for everything, so it does not necessarily need a structure. but I'll add it, I don't know why I removed it

ElliottKasoar commented on docs/source/user_guide/calculations.rst:
```diff
@@ -1,9 +1,18 @@
+.. note::
+   Any format that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read is a valid structure file for a calculation.
+
+The model file determines the specific MLIP to be used. I can be a local file or a URI to a file to download. In these examples, it is assumed to be a local file located at `path/to/model`.
```
```suggestion
The model file determines the specific MLIP to be used. It can be a local file or a URI to a file to download. In these examples, it is assumed to be a local file located at `path/to/model`.
```

ElliottKasoar commented on docs/source/user_guide/calculations.rst:
```diff
@@ -18,24 +27,82 @@ Below is a usage example with the minimum required parameters. These parameters
+    submit(SinglePointCalculation, **inputs)
+
+
+Or they can be passed as a config file. The config file has to be structured as it would be for a janus calculation (refer to janus documentation) and passed as an AiiDA data type itself.
```
Maybe this should link to janus and/or the docs? I'll look for the rst syntax to suggest but you may know immediately?

ElliottKasoar commented on docs/source/user_guide/calculations.rst:
```diff
@@ -18,24 +27,82 @@ Below is a usage example with the minimum required parameters. These parameters
+    submit(SinglePointCalculation, **inputs)
+
+
+Or they can be passed as a config file. The config file has to be structured as it would be for a janus calculation (refer to `janus documentation <https://stfc.github.io/janus-core/apidoc/janus_core.html>`_ ) and passed as an AiiDA data type itself.
```
```suggestion
Or they can be passed as a config file. The config file has to be structured as it would be for a janus calculation (refer to `janus documentation <https://stfc.github.io/janus-core/>`_ ) and passed as an AiiDA data type itself.
```
Optional but I'm thinking the main page might be better to link, since the API docs aren't the most user friendly starting point

federicazanca replied:
yeah I know, I put api cause at the moment the stuff is mostly there

ElliottKasoar reviewed (approved):
Optional suggestion, but otherwise I think this looks good.

We'll revisit a lot of the option names, config etc. anyway at some point.

alinelena commented on aiida_mlip/calculations/geomopt.py:
```diff
@@ -44,49 +44,44 @@ def define(cls, spec: CalcJobProcessSpec) -> None:
+            default=lambda: Str(cls.DEFAULT_TRAJ_FILE),
+            help="Path to save optimisation frames to",
         )
         spec.input(
```
no default?

federicazanca replied:
I'll write something in the documentation to exaplain how it works at the moment and open another issue to fix some things with the output in the future

ElliottKasoar commented on docs/source/user_guide/calculations.rst:
```diff
@@ -18,24 +27,102 @@ Below is a usage example with the minimum required parameters. These parameters
 
 Refer to the API documentation for additional parameters that can be passed.
+Some parameters are not required and don't have a default value set in aiida-mlip. In that case the default values will be the same as `janus <https://stfc.github.io/janus-core/>`_
+The only default parameters defined in aiida-mlip are the names of the input and output files, are they do not affect the results of the calculation itself, and are needed in AiiDA to parse the results.
```
```suggestion
The only default parameters defined in aiida-mlip are the names of the input and output files, as they do not affect the results of the calculation itself, and are needed in AiiDA to parse the results.
```

ElliottKasoar commented on aiida_mlip/parsers/base_parser.py:
```diff
@@ -0,0 +1,86 @@
+            An exit code.
+        """
+        output_filename = self.node.get_option("output_filename")
+        logoutput = (self.node.inputs.log_filename).value
+
+        # Check that folder content is as expected
+        files_retrieved = self.retrieved.list_object_names()
+
+        files_expected = {output_filename, logoutput}
+        if not files_expected.issubset(files_retrieved):
+            self.logger.error(
+                f"Found files '{files_retrieved}', expected to find '{files_expected}'"
+            )
+            return self.exit_codes.ERROR_MISSING_OUTPUT_FILES
+
+        # Add output file to the outputs
+
+        with (
+            self.retrieved.open(logoutput, "rb") as log,
```
```suggestion
        log_output = (self.node.inputs.log_filename).value

        # Check that folder content is as expected
        files_retrieved = self.retrieved.list_object_names()

        files_expected = {output_filename, log_output}
        if not files_expected.issubset(files_retrieved):
            self.logger.error(
                f"Found files '{files_retrieved}', expected to find '{files_expected}'"
            )
            return self.exit_codes.ERROR_MISSING_OUTPUT_FILES

        # Add output file to the outputs

        with (
            self.retrieved.open(log_output, "rb") as log,
```

ElliottKasoar commented on aiida_mlip/parsers/sp_parser.py:
```diff
@@ -93,16 +74,19 @@ def parse(self, **kwargs) -> int:
+        if exit_code != ExitCode(0):
+            return exit_code
+
+        xyzoutput = (self.node.inputs.out).value
```
I can't suggest changes to all the instances, but `xyz_output` would be nicer, as it's quite hard to read as is.
